subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/MutableTransformExecution.java

```diff
@@ -42,14 +46,23 @@ public MutableTransformExecution(
             transform, inputArtifact, dependencies, subject,
             transformExecutionListener, buildOperationExecutor, fileCollectionFactory, inputFingerprinter, workspaceServices
         );
+        this.rootProjectLocation = producerProject.getRootDir().getAbsolutePath() + File.separator;
     }
 
     @Override
     public Identity identify(Map<String, ValueSnapshot> identityInputs, Map<String, CurrentFileCollectionFingerprint> identityFileInputs) {
         return new MutableTransformWorkspaceIdentity(
-            inputArtifact.getAbsolutePath(),
+            normalizeAbsolutePath(inputArtifact.getAbsolutePath()),
             identityInputs.get(AbstractTransformExecution.SECONDARY_INPUTS_HASH_PROPERTY_NAME),
             identityFileInputs.get(AbstractTransformExecution.DEPENDENCIES_PROPERTY_NAME).getHash()
         );
     }
+
+    private String normalizeAbsolutePath(String path) {
+        // We try to normalize the absolute path, so the workspace id is stable between machines for cacheable transforms.
+        if (path.startsWith(rootProjectLocation)) {
+            return path.substring(rootProjectLocation.length());
+        }
+        return path;
+    }
```

Comment:
```suggestion
    Path rootProjectLocation;

    private String normalizeArtifactPath(File inputArtifact) {
        // We try to normalize the absolute path, so the workspace id is stable between machines for cacheable transforms.
        Path path = inputArtifact.toPath().toAbsolutePath();
        if (path.startsWith(rootProjectLocation)) {
            return path.relativize(rootProjectLocation).toString();
        }
        return path.toString();
    }
```


What are your thoughts on using some of the `Path` methods to do this instead of simple string operations?

Reply:
I think I tried it and it was too slow performance wise. I can try again and see what I end up with. The string comparison is really fast and works because all the paths involved here are created by Gradle and users can't do "funky" stuff there.

Reply:
Though, if Gradle is generating all of these paths, could we generate them originally with a normalized path?

Reply:
Well, we mostly want the absolute paths to actually read the files from disk...

Reply:
Ahh, yeah.

Reply:
Let's see what the performance test says: https://builds.gradle.org/buildConfiguration/Gradle_Master_Util_Performance_AdHocPerformanceScenarioLinuxAmd64/67133370?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true

Reply:
Seem like there is about 3% performance difference between the two approaches (the String based one being faster), so I'd keep it.